d2ir/compile.go

```diff
@@ -0,0 +1,251 @@
+package d2ir
+
+import (
+	"oss.terrastruct.com/d2/d2ast"
+	"oss.terrastruct.com/d2/d2parser"
+)
+
+type compiler struct {
+	err d2parser.ParseError
+}
+
+func (c *compiler) errorf(n d2ast.Node, f string, v ...interface{}) {
+	c.err.Errors = append(c.err.Errors, d2parser.Errorf(n, f, v...).(d2ast.Error))
+}
+
+func Compile(ast *d2ast.Map) (*Map, error) {
+	c := &compiler{}
+	m := &Map{}
+	m.initRoot()
+	m.parent.(*Field).References[0].Context.Scope = ast
+	c.compileMap(m, ast)
+	c.compileScenarios(m)
+	c.compileSteps(m)
+	if !c.err.Empty() {
+		return nil, c.err
+	}
+	return m, nil
+}
+
+func (c *compiler) compileScenarios(m *Map) {
+	scenariosf := m.GetField("scenarios")
+	if scenariosf == nil {
+		return
+	}
+	scenarios := scenariosf.Map()
+	if scenarios == nil {
+		return
+	}
+
+	for _, sf := range scenarios.Fields {
+		if sf.Map() == nil {
+			continue
+		}
+		base := m.CopyBase(sf)
+		sf.Composite = Overlay(base, sf.Map())
+		c.compileScenarios(sf.Map())
+		c.compileSteps(sf.Map())
+	}
+}
+
+func (c *compiler) compileSteps(m *Map) {
+	stepsf := m.GetField("steps")
+	if stepsf == nil {
+		return
+	}
+	steps := stepsf.Map()
+	if steps == nil {
+		return
+	}
+	for i, sf := range steps.Fields {
+		if sf.Map() == nil {
+			continue
+		}
+
+		var base *Map
+		if i == 0 {
+			base = m.CopyBase(sf)
+		} else {
+			base = steps.Fields[i-1].Map().CopyBase(sf)
+		}
+		sf.Composite = Overlay(base, sf.Map())
+		c.compileScenarios(sf.Map())
+		c.compileSteps(sf.Map())
+	}
+}
+
+func (c *compiler) compileMap(dst *Map, ast *d2ast.Map) {
+	for _, n := range ast.Nodes {
+		switch {
+		case n.MapKey != nil:
+			c.compileKey(&RefContext{
+				Key:      n.MapKey,
+				Scope:    ast,
+				ScopeMap: dst,
+			})
+		case n.Substitution != nil:
+			panic("TODO")
```

Comment:
It can substitutions are parsed though I can't imagine someone would do a `...$dsmsa` unless they were testing the syntax. I'll just comment out the panic for now.